### PR TITLE
Implementation of parsing new binary snapshot format

### LINKED ISSUE #7188: Follow up with implementation of parsing to improved binary snapshot in odsp driver
Spo is having some improvements in binary snapshot for trees/latest in the odsp driver. So follow up with the implementation of parsing in driver once it is done in the server.

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/7188
Odsp server have implementation of serializing the latest snapshot in binary format that changed. So we need to change the corresponding parser at client side.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/WriteBufferUtils.ts [hunk 1]
```diff
@@ -75,6 +84,34 @@ const binaryBytesToCodeMap = {
     8: 16,
 };
 
+/**
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/WriteBufferUtils.ts:92]: Added link to the file where these codes are specified. Also opening an issue for npm package suggestion.

[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/WriteBufferUtils.ts:92]: https://github.com/microsoft/FluidFramework/issues/8052

**Comment by jatgarg** [comment]
Ping @vladsud

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 2]
```diff
@@ -17,69 +19,23 @@ import {
...
-    for (const [idIndexNode, blobNode] of node.iteratePairs()) {
-        assertNumberInstance(idIndexNode, "Blob index should be a number");
-        assertBlobCoreInstance(blobNode, "Blob content should be of type blob");
-        const blobId = dictionary[idIndexNode];
-        assert(blobId !== undefined, 0x214 /* "blob id should be present" */);
-        blobs.set(blobId, blobNode.arrayBuffer);
+    for (let count = 0; count < node.length; ++count) {
+        const blob = node.getNode(count);
+        const records = getAndValidateNodeProps(blob, ["id", "data"]);
+        assertBlobCoreInstance(records.data, "data should be of BlobCore type");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:36]: Do we have an issue tracking tagging assertBlobCoreInstance and similar APIs to remove strings from final bundles?

**Comment by vladsud** [comment]
Note size increase:

<html>
<body>


odspDriver.js | 182.73 KB | 184.64 KB | ⯅ +1.91 KB
-- | -- | -- | --
odspPrefetchSnapshot.js | 39.38 KB | 41.29 KB | ⯅ +1.91 KB


</body>
</html>

I do not know how substantial it is to shipping scenarios, but if there is a way for odspPrefetchSnapshot.js flow not to have experimental path (i.e. V2) - that might be good starting point, as that's the only thing that goes into apps, the rest is dynamically delivered (not bundled with apps), so there is a bit less pressure RE size growth.

Curious, how much of that is caused by assert strings, and if there is anything that is getting into bundles that is not shipping code (like write code)? Worth examining bundles and their composition to make sure we do the best RE size increase.

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:36]: Yes it is here: https://github.com/microsoft/FluidFramework/issues/7185

**Comment by jatgarg** [comment]
> Note size increase:
> 
> odspDriver.js	182.73 KB	184.64 KB	⯅ +1.91 KB
> odspPrefetchSnapshot.js	39.38 KB	41.29 KB	⯅ +1.91 KB
> I do not know how substantial it is to shipping scenarios, but if there is a way for odspPrefetchSnapshot.js flow not to have experimental path (i.e. V2) - that might be good starting point, as that's the only thing that goes into apps, the rest is dynamically delivered (not bundled with apps), so there is a bit less pressure RE size growth.
> 
> Curious, how much of that is caused by assert strings, and if there is anything that is getting into bundles that is not shipping code (like write code)? Worth examining bundles and their composition to make sure we do the best RE size increase.

This new api will be going away soon. So I don't think it will 1KB is a big deal as the assert strings will also go away. 
If you are saying that we do not want the new format for prefetchSnapshot api, then that is separate thing and we can see if we can achieve that as the code flow will not allow that currently.

**Pull request merged**